### PR TITLE
dhcpv6_client: refactor to use `event_timeout` for non-sock timeouts

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Not sure, why I did not do that in the first place (maybe `event_timeout` did not exist back then, I seemed to basically have mirrored its behavior). As an added bonus, this makes the migration to `ztimer` really easy, which I also did in this PR.

### Testing procedure
`make -C tests/gnrc_dhcpv6_client flash -j test` still passes, with and without `USEMODULE=ztimer`

### Issues/PRs references
~~Requires https://github.com/RIOT-OS/RIOT/pull/16667 for the test to pass with `ztimer`.~~ (merged)

#16661 conflicts with this PR, but IMHO #16661 should be merged first, to have the issue fixed in master.

### CODE REVIEW

**Comment by miri64** [comment]
Some of the changes (such as the general microsecond -> millisecond conversion) are mainly to prepare for a port to `sock_async` + `event_timeout` for retransmissions (so `dhcpv6_client` and `dhcpv6_relay` can run in the same thread without the client blocking on `sock_udp_recv`), but I introduced it here already to not have confusing conversions for `ztimer`.

File: sys/net/application_layer/dhcpv6/client.c
Comment (kfessel) [inline sys/net/application_layer/dhcpv6/client.c:210]: reads like this should use `MS_PER_CS`
```suggestion
    return (uint32_t)(ztimer_now(ZTIMER_MSEC) / MS_PER_CS);
```

File: sys/net/application_layer/dhcpv6/client.c
Comment (kfessel) [inline sys/net/application_layer/dhcpv6/client.c:219]: needs a factor of 1/1000?
```suggestion
    return (uint32_t)(ztimer_now(ZTIMER_MSEC) / MS_PER_SEC));
```

**Review by kfessel — COMMENTED** [review]
some time unit typos

File: sys/net/application_layer/dhcpv6/client.c
Comment (miri64) [reply sys/net/application_layer/dhcpv6/client.c:219]: I think that was supposed to be
```suggestion
    return (uint32_t)ztimer_now(ZTIMER_SEC);
```

**Comment by miri64** [comment]
Rebased and squashed to current master.

File: sys/net/application_layer/dhcpv6/client.c
Comment (kfessel) [inline sys/net/application_layer/dhcpv6/client.c:508]: this and  l.: 842 seem to be the only lines that access `rebind_time`, they cancel each other  `+t2` and  `-t2`

at a first glance the `-t2` in  l.: 842 may be wrong (but might have been working around: failing renew since `mrd` seems to be used to evaluate if a renew is atempted '-t2' make this only be done if the renew is run in the same second as the "_schedule_t2").

**Review by kfessel — COMMENTED** [review]
A thought about rebind time an its offset by t2.

File: sys/net/application_layer/dhcpv6/client.c
Comment (miri64) [reply sys/net/application_layer/dhcpv6/client.c:508]: This PR aims not to change the logic of the DHCPv6 implementation, but only refactors, so even if it is wrong, that change would belong into a different PR. Hovewer, L. 842 calculates the MRD for a RENEW transmission (please refer to the RFC to figure out what this means exactly, the implementation was a few years ago, I would need to reread it myself, to explain it properly myself; all the more reason why I don't want to change logic in this PR ... ;-)). This line only reschedules the next T2 event which was supplied in a REPLY message by the IA_PD option (or IA_NA option, once #16228 is in), so I don't see how the two operations would be related. I guess in L. 842 we want some kind of offset to T2, as for RENEW the MRD shall be ["Remaining time until earliest T2"](https://datatracker.ietf.org/doc/html/rfc8415#section-18.2.4).

Concerns about things happening at the same time don't make sense, since the client runs in a single thread ;-).

Comment (kfessel) [reply sys/net/application_layer/dhcpv6/client.c:508]: Something is wrong with the mrd calculation in the case of "renew" and "rebind" (request does not calculate an mrd).

(rebind will never happen, renew may run longer than t2 allows)

The error does not influence how the system works if the first renew is successful -> dificult to see in the wild

I only looked at these lines because i checked on the xtimer ztimer conversion.

The error(s) was in there before.

everything done in this PR is working as expected

**Review by kfessel — APPROVED** [review]
I ran this on native following the "Testing procedure". :heavy_check_mark: 
Using short T1 and T2, renew was successful for xtimer as well as for ztimer. :heavy_check_mark: 

I like the changes of this patch. :heavy_check_mark: 

Maybe i would only use msec (one ztimer to rule them all) :question:

File: sys/net/application_layer/dhcpv6/client.c
Comment (miri64) [inline sys/net/application_layer/dhcpv6/client.c:960]: > Maybe i would like the use of just msec (one ztimer to rule them all, this could also avoid some jitter but dhcpv6 seems to be tolerant to that) question

Main reason for also using `ZTIMER_SEC` is this line. I really do not want to cap of a potential overflow here. Since this function is used to set the T1 and T2 timeouts (both come over the wire) it is not guaranteed that they will be e.g. `UINT32_MAX - 1`.

Comment (kfessel) [reply sys/net/application_layer/dhcpv6/client.c:960]: This may still be problematic since gnrc_dhcp `dhcpv6_client_conf_prefix` and nib `_nib_pl_add` have a max addr lifetime of uint32_max-1 msecs or infite
which will make the addr last shorter that its T1 and T2 timeout

Comment (miri64) [reply sys/net/application_layer/dhcpv6/client.c:960]: Sure, but nothing prevents us to port that to `ZTIMER_SEC` in the future as well ;-). Currently the same problem exists in master, since the `xtimer` of DHCPv6 uses 64-bit scale.

**Comment by miri64** [comment]
Now this PR only needs an ACK by a maintainer.

File: sys/net/application_layer/dhcpv6/client.c
Comment (maribu) [inline sys/net/application_layer/dhcpv6/client.c:223]: There is a potential footgun with using both timestamps from ZTIMER_SEC and ZTIMER_MSEC, as both could be provided by different hardware. One could (or at least I can see myself doing so) assume that the timestamps can be converted to each other.

I don't see this incorrect assumption anywhere here in the code, but I'm not really familiar with it.

**Review by maribu — APPROVED** [review]
ACK. Change looks sane to me and I trust @kfessel testing. Please confirm that the potential issue I pointed out inline is indeed not present before hitting merge.

File: sys/net/application_layer/dhcpv6/client.c
Comment (miri64) [reply sys/net/application_layer/dhcpv6/client.c:223]: Yes, the timestamp should be used independent from each other and are only used for the respective timeouts using the different scales.

Comment (kfessel) [reply sys/net/application_layer/dhcpv6/client.c:223]: there are multiple potential foot guns, but none of them is worse than it was before with just xtimer, especialiy within reasnonable limit of leasetime and T1 and T2 I would advice against giving longer than 1 month leases

Comment (kfessel) [reply sys/net/application_layer/dhcpv6/client.c:223]: The drift pontential of msec vs sec is low enough for T1 and T2 application,

The problem left is one might uses SEC as some kind of RTC (which should not be done), combine that with DHCP and some NTP after getting the dhcp lease and you get a time scramble, this is a problem of the ztimer and not of this DHCP.

**Comment by kfessel** [comment]
i also ran it on the board ethernet port of a nucleo-f767zi (the biggest dificulty was to find the place where "ethos" was configured)

it gets its  lease and renews them (short duration)

**Pull request merged**
